File: reader.py
#!/bin/bash
from typing import Optional
import socket
import urllib.request

sep : list = list(map(lambda x:  bytes(x, 'utf-8'), "--boundarydonotcross\r\n"))
sep_len : int = len(sep)


def next_boundry(socket: socket) -> list:
    bytes = []

    #for i in range (0,335):
    while True:
        data = s.recv(1)
        bytes.append(data)

        if len(bytes) >= len(sep) and bytes[-sep_len:] == sep:
            break


    # return all but seperator
    #print(bytes)
    return bytes[0:-sep_len]


def remove_header(data: list) -> list:
    for i in range(len(data)):
        if data[i:i+4] == [b'\xff', b'\xd8', b'\xff', b'\xe0']:
            return data[i:]
    print("error")
    exit(1)



if __name__ == "__main__":
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    #s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    #s._fileobject.default_bufsize = 0
    s.connect(("127.0.0.1", 8080))
    request = "GET /?action=stream HTTP/1.1\r\nHost:%s\r\n\r\n"
    s.send(request.encode())

    # HTTP Header
    next_boundry(s)

    # first frame
    frame = next_boundry(s)
    frame = remove_header(frame)
    text = b''.join(frame)
    f = open('out1.jpg', 'wb')
    f.write(text)

    # next frame
    frame = next_boundry(s)
    frame = remove_header(frame)
    text = b''.join(frame)
    f = open('out2.jpg', 'wb')
    f.write(text)

File: reader.go
//https://go101.org/article/channel-use-cases.html
package main

import (
	"bufio"
	"bytes"
	"net"
	"os"
	"sync"
	"time"
)

type Frame struct {
	header []byte
	body   []byte
}

var delim = []byte("--boundarydonotcross\r\n")
var frame_start = []byte("\xff\xd8\xff\xe0")

func read_til_boundry(r *bufio.Reader) (data []byte) {
	for {
		buffer, err := r.ReadString('\n')

		if err != nil {
			panic("ERRROR")
		}

		data = append(data, []byte(buffer)...)

		if bytes.HasSuffix(data, delim) {
			return data[:len(data)-len(delim)]
		}
	}
}

func parse_frame(data []byte) (frame Frame) {
	for i := 0; i < len(data); i++ {
		if bytes.Compare(data[i:i+4], frame_start) == 0 {
			return Frame{data[:i], data[i:]}
		}
	}
	panic("invalid frame")
}

func start_socket(port string) (reader *bufio.Reader) {
	conn, err := net.Dial("tcp", "127.0.0.1:"+port)
	if err != nil {
		panic("Socket error")
	}
	conn.Write([]byte("GET /?action=stream HTTP/1.1\r\nHost:%s\r\n\r\n"))

	return bufio.NewReader(conn)

}

func source(reader *bufio.Reader) chan Frame {
	var source = make(chan Frame)
	go func() {
		//header
		_ = read_til_boundry(reader)
		for {
			var data = read_til_boundry(reader)
			var frame = parse_frame(data)
			//os.WriteFile("./out_" +port+ ".jpg", frame, 0644)
			source <- frame
		}
	}()

	return source
}

func sink(sources []chan Frame) {
	go func() {
		for {
			for _, source := range sources {
				var frame = <-source
				os.WriteFile("./out_.jpg", frame.body, 0644)
				time.Sleep(250 * time.Millisecond)
			}
		}
	}()
}

func main() {
	var args = os.Args[1:]
	var wg sync.WaitGroup

	var channels []chan Frame

	for _, port := range args {
		wg.Add(1)
		var reader = start_socket(port)
		var channel = source(reader)
		channels = append(channels, channel)
	}

	wg.Add(1)
	sink(channels)

	wg.Wait()
}
